lstchain/reco/reconstructor.py

```diff
@@ -0,0 +1,944 @@
+            if max_mu < 120:
+                kmax = int(np.ceil(1.34 * (max_mu-20) + 45))
+            else:
+                kmax = int(np.ceil(1.096 * max_mu + 47.8))
+        if kmin < 0:
+            kmin = 0
+        if kmax > self.n_peaks:
+            kmax = int(self.n_peaks)
+            logger.warning("kmax forced to %s", kmax)
+
+        photo_peaks = np.arange(kmin, kmax, dtype=int)
+        crosstalk_factor = photo_peaks[..., None]*self.crosstalk[mask_LL]
+
+        # Compute the Poisson term in the pixel likelihood for
+        # low luminosity pixels
+        mu_plus_crosstalk = mu[mask_LL] + crosstalk_factor
+        log_mu_plus_crosstalk = ne.evaluate("log(mu_plus_crosstalk)",
+                                            local_dict=dict(mu_plus_crosstalk=mu_plus_crosstalk))
+        log_mu_plus_crosstalk = ((photo_peaks - 1)
+                                 * log_mu_plus_crosstalk.T).T
+        log_poisson = (log_mu[mask_LL]
+                       - self.log_factorial[kmin:kmax][..., None]
+                       - mu_plus_crosstalk
+                       + log_mu_plus_crosstalk)
+
+        # Compute the Gaussian term in the pixel likelihood for
+        # low luminosity pixels
+        signal = self.data
+
+        if self.use_weight:
+            weight = 1.0+(signal/np.max(signal))
+        else:
+            weight = np.ones(signal.shape)
+
+        mean = (photo_peaks
+                * templates[mask_LL][..., None])
+        sigma_n = (photo_peaks
+                   * ((self.sigma_s[mask_LL][..., None]*templates[mask_LL])**2)[..., None])
+        sigma_n = (self.error[mask_LL]**2)[..., None] + sigma_n
+        sigma_n = ne.evaluate("sqrt(sigma_n)", local_dict=dict(sigma_n=sigma_n))
+        log_gauss = log_gaussian(signal[mask_LL][..., None], mean, sigma_n)
+
+        # Compute the pixel likelihood using a Gaussian approximation for
+        # high luminosity pixels
+        if np.any(~mask_LL):
+            mu_hat = ((mu[~mask_LL] / (1-self.crosstalk[~mask_LL]))[..., None]
+                      * templates[~mask_LL])
+            sigma_hat = (((mu[~mask_LL]
+                           / np.power(1-self.crosstalk[~mask_LL], 3))[..., None]
+                         * templates[~mask_LL]**2))
+            sigma_hat = np.sqrt((self.error[~mask_LL]**2) + sigma_hat)
+
+            if self.use_weight:
+                log_pixel_pdf_HL = weight[~mask_LL] * log_gaussian(signal[~mask_LL], mu_hat, sigma_hat)
+            else:
+                log_pixel_pdf_HL = log_gaussian(signal[~mask_LL], mu_hat, sigma_hat)
+            n_points_HL = np.sum(weight[~mask_LL])
+        else:
+            log_pixel_pdf_HL, n_points_HL = np.asarray([0]), 0
+
+        log_poisson = np.expand_dims(log_poisson.T, axis=1)
+        log_pixel_pdf_LL = ne.evaluate("log_poisson + log_gauss",
+                                       local_dict=dict(log_poisson=log_poisson, log_gauss=log_gauss))
+        if self.use_weight:
```

Comment:
You have many `if self.use_weight` statements here. It looks like it would be much better to just have two functions for the weighted and unweighted cases.